lib/duniverse.ml

```diff
@@ -19,34 +23,37 @@ module Deps = struct
 
   module Source = struct
     module Package = struct
-      type t = { opam : Opam.t; upstream : string; ref : string }
+      type t = { opam : Opam.t; upstream : string; ref : Git.Ref.t }
 
       let equal t t' =
         Opam.equal t.opam t'.opam
         && String.equal t.upstream t'.upstream
-        && String.equal t.ref t'.ref
+        && Git.Ref.equal t.ref t'.ref
 
       let raw_pp fmt { opam; upstream; ref } =
         Format.fprintf fmt "@[<hov 2>{ opam = %a;@ upstream = %S;@ ref = %S }@]" Opam.raw_pp opam
           upstream ref
     end
 
-    type t = {
+    type 'ref t = {
       dir : string;
       upstream : string;
-      ref : string; [@default "master"] [@sexp_drop_default]
+      ref : 'ref;
       provided_packages : Opam.t list [@default []] [@sexp_drop_default]
     }
     [@@deriving sexp]
 
-    let equal t t' =
-      String.equal t.dir t'.dir && String.equal t.upstream t'.upstream && String.equal t.ref t'.ref
+    let equal equal_ref t t' =
+      String.equal t.dir t'.dir
```

Comment:
I suggest using pattern matching here to make sure that it breaks when more fields are added.

Reply:
Yes good idea!